### PR TITLE
Matmul_nbits kernel for mlas sqnbits to support Fp16 inputs

### PR DESCRIPTION
**Pull request by liqunfu**
### Description
There are 2 ways to support Fp16:

1. At low level to use fp16 specific instructions:
https://github.com/microsoft/onnxruntime/pull/21760
This requires instruction support and, in case of avx, halfgemm support which is missing.
[this PR](https://github.com/microsoft/onnxruntime/pull/21760)  can complete after blksum compute is moved away from sgemm (https://github.com/microsoft/onnxruntime/tree/liqun/blksum)

2. Convert Fp16 to Fp32 and use the mlas Fp32 version.
This PR is to implement the second approach.  

### Motivation and Context
support Fp16

### CODE REVIEW

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2.cpp [hunk 1]
```diff
@@ -29,6 +29,47 @@ Module Name:
 #include "sqnbitgemm_m1_sym_kernel_avx2_int8_blklen32.h"
 #include "sqnbitgemm_m1_sym_kernel_avx2_int8_blklen64.h"
 
+void ConvertFp16ToFp32Avx(const MLAS_FP16* src_fp16, float* dst_fp32, uint64_t size)
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2.cpp:32]: there is an existing implementation for this: MlasConvertHalfToFloatBuffer in MLAS.h

File: onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc [hunk 2]
```diff
@@ -79,6 +79,7 @@
   return true;
 }
 
+template <typename AType>
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc:82]: >AType [](http://example.com/codeflow?start=19&length=5)

Do you need a template for whole class? The data type of A and scale can be retrieved from Tensor A and Tensor Scale

File: onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2.cpp (hunk 1 above)
Comment (liqunfu) [reply onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx2.cpp]: consolidated.

File: onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc (hunk 2 above)
Comment (liqunfu) [reply onnxruntime/contrib_ops/cpu/quantization/matmul_nbits.cc]: removed the template.

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**